File: Binary Tree Postorder Traversal/Binary Tree Postorder Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        vector<int> res;
        stack<TreeNode *> stack;
        TreeNode *node = root;
        TreeNode *prev = nullptr;
        while (false == stack.empty() || nullptr != node) {
            if (nullptr != node) {
                stack.push(node);
                node = node->left;
            } else {
                TreeNode *peek = stack.top();
                if (nullptr != peek->right && prev != peek->right) {
                    node = peek->right;
                } else {
                    res.push_back(peek->val);
                    stack.pop();
                    prev = peek;
                }
            }
        }
        return res;
    }
};

File: Binary Tree Postorder Traversal/Binary Tree Postorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList();
        TreeNode prev = null;
        Stack<TreeNode> stack = new Stack();
        while (false == stack.isEmpty() || null != root) {
            if (null != root) {
                stack.push(root);
                root = root.left;
            } else {
                TreeNode peek = stack.peek();
                if (null != peek.right && prev != peek.right) {
                    root = peek.right;
                } else {
                    res.add(peek.val);
                    stack.pop();
                    prev = peek;
                }
            }
        }
        return res;
    }
}